dagstore.go

```diff
@@ -0,0 +1,609 @@
+type Task struct {
+	Op    OpType
+	Shard *Shard
+	Resp  chan ShardResult
+	Index carindex.Index
+	Error error
+}
+
+// ShardResult encapsulates a result from an asynchronous operation.
+type ShardResult struct {
+	Key      shard.Key
+	Error    error
+	Accessor *ShardAccessor
+}
+
+type Config struct {
+	// ScrapRoot is the path to the scratch space, where local copies of
+	// remote mounts are saved.
+	ScratchSpaceDir string
+
+	// IndexDir is the path where indices are stored.
+	IndexDir string
+
+	// Datastore is the datastore where shard state will be persisted.
+	Datastore ds.Datastore
+
+	// MountTypes are the recognized mount types, bound to their corresponding
+	// URL schemes.
+	MountTypes map[string]mount.Type
+}
+
+// NewDAGStore constructs a new DAG store with the supplied configuration.
+func NewDAGStore(cfg Config) (*DAGStore, error) {
+	// validate and manage scratch root directory.
+	if cfg.ScratchSpaceDir == "" {
+		return nil, fmt.Errorf("missing scratch area root path")
+	}
+	if err := ensureDir(cfg.ScratchSpaceDir); err != nil {
+		return nil, fmt.Errorf("failed to create scratch root dir: %w", err)
+	}
+
+	// instantiate the index repo.
+	var indices index.FullIndexRepo
+	if cfg.IndexDir == "" {
+		log.Info("using in-memory index store")
+		indices = index.NewMemoryRepo()
+	} else {
+		err := ensureDir(cfg.IndexDir)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create index root dir: %w", err)
+		}
+		indices, err = index.NewFSRepo(cfg.IndexDir)
+		if err != nil {
+			return nil, fmt.Errorf("failed to instantiate full index repo: %w", err)
+		}
+	}
+
+	// handle the datastore.
+	if cfg.Datastore == nil {
+		log.Warnf("no datastore provided; falling back to in-mem datastore; shard state will not survive restarts")
+		cfg.Datastore = ds.NewMapDatastore()
+	}
+
+	// create the registry and register all mount types.
+	mounts := mount.NewRegistry()
+	for scheme, typ := range cfg.MountTypes {
+		if err := mounts.Register(scheme, typ); err != nil {
+			return nil, fmt.Errorf("failed to register mount factory: %w", err)
+		}
+	}
+
+	// TODO: recover persisted shard state from the Datastore.
+	// TODO: instantiate scratch space with persisted paths.
+	// scratch, err := NewScratchSpace(cfg.ScratchSpaceDir, map[shard.Key]string{}, strconv.Itoa(int(time.Now().Unix())))
+	// if err != nil {
+	// 	return nil, fmt.Errorf("failed to create scratch space: %w", err)
+	// }
+
+	ctx, cancel := context.WithCancel(context.Background())
+	dagst := &DAGStore{
+		mounts:       mounts,
+		config:       cfg,
+		indices:      indices,
+		shards:       make(map[shard.Key]*Shard),
+		externalCh:   make(chan *Task, 128), // len=128, concurrent external tasks that can be queued up before putting backpressure.
+		internalCh:   make(chan *Task, 1),   // len=1, because eventloop will only ever stage another internal event.
+		completionCh: make(chan *Task, 64),  // len=64, hitting this limit will just make async tasks wait.
```

Comment:
We don't _need_ it right now, but it separates concerns nicely, and makes the code more future proof in case we do need to process tasks differently going forward.